guy/my_ps/proc_info.c

```diff
@@ -0,0 +1,131 @@
+ * dereference_symlinks [IN] -> If true, read the file the symlink point to rather than the link file itself.
+* RETURN -> Number of bytes read.
+ */
+size_t read_proc_fs_file(const char *path, char *file_contents, bool dereference_symlinks) {
+    size_t bytes_read = 0;
+    struct stat file_stat;
+
+    if (fstatat(0, path, &file_stat, AT_SYMLINK_NOFOLLOW)) {
+        if (DEBUG) {
+            printf("Couldn't stat \"%s\". Check it exists and has the right permissions set.\n", path);
+        }
+
+        file_contents = NULL;
+        return 0;
+    }
+    // If told not to dereference links, check if the file is a link and if it is return its target.
+    if (!dereference_symlinks && S_ISLNK(file_stat.st_mode)) {
+        bytes_read = readlink(path, file_contents, PATH_MAX - 1);
+        file_contents[bytes_read] = '\0';
+        return bytes_read;
+    }
+
+    FILE *fd = fopen(path, "r");
+    if (NULL == fd) {
+        if (DEBUG) {
+            printf("An Error has occurred while opening \"%s\".\n", path);
+        }
+    } else {
+        // Since it turns out we cant use lseek or stat to find out file size, we will to read
+        //bytes_read = fread((void*)file_contents, MAX_PROC_FS_FILE_SIZE, 1, fd);
+        for (bytes_read = 0;
+             bytes_read < MAX_PROC_FS_FILE_SIZE && EOF != (file_contents[bytes_read] = fgetc(fd)); bytes_read++);
+    }
+    fclose(fd);
+    return bytes_read;
+}
+
+/*
+ * CANT HANDLE LINKS!!!
+ * Copies the file's content in a buffer. Built to work with the proc fs.
+ * proc_id [IN] -> Process ID.
+ * file_name [IN] -> The name of the file to read. (relative to a /proc/x/ dir.
+ * file_contents [OUT] -> buffer to which we copy the file's contents.
+* RETURN -> Number of bytes read.
+ */
+size_t get_proc_file_contents(const unsigned int proc_id, const char *file_name, char *file_contents) {
+    size_t bytes_read = 0;
+    char *file_path;
+    file_path = (char *) malloc(PATH_MAX);
+    sprintf(file_path, "/proc/%d/%s", proc_id, file_name);
+    bytes_read = read_proc_fs_file(file_path, file_contents, false);
+    if (!bytes_read) {
+        if (DEBUG) {
+            printf("An Error has occurred while reading file \"%s\" of proc %d.\n", file_path, proc_id);
+        }
+    }
+    free((void *) file_path);
+    return bytes_read;
+}
+
+/*
+ * Initialize a process info struct and load it with relevant data.
+ *
+ * *pi [OUT] -> Pointer to a struct to initialize.
+ * pid [IN] -> The PID of process.
+ * RETURNS -> Error Code.
+ */
+unsigned int init_proc_info(proc_info_t *pi, const unsigned int pid) {
+    size_t bytes_read = 0;
+    unsigned int i = 0;
+    pi->id = pid;
+
+    // Allocate memory buffers for name, command line and exec path.
+    pi->name = (char *) malloc(MAX_PROC_FS_FILE_SIZE);
+    pi->cmd_line = (char *) malloc(MAX_PROC_FS_FILE_SIZE);
+    pi->exe_path = (char *) malloc(MAX_PROC_FS_FILE_SIZE);
+    if (NULL == pi->name && NULL == pi->cmd_line && NULL == pi->exe_path) {
```

Comment:
and what would happen if only one of the allocations failed?

Reply:
Must've confused && and ||. Fixed

Reply:
assume only one of the allocations failed, don't you need to free the others?